builtin/v9/migration/test/migration_test.go

```diff
@@ -0,0 +1,99 @@
+package migration
+
+import (
+	"context"
+	"testing"
+
+	"github.com/filecoin-project/go-address"
+	"github.com/filecoin-project/go-state-types/manifest"
+	"github.com/ipfs/go-cid"
+
+	"github.com/filecoin-project/go-state-types/builtin/v9/migration"
+
+	"github.com/filecoin-project/go-state-types/builtin"
+	system9 "github.com/filecoin-project/go-state-types/builtin/v9/system"
+	"github.com/filecoin-project/go-state-types/builtin/v9/util/adt"
+	cbor "github.com/ipfs/go-ipld-cbor"
+	"github.com/stretchr/testify/require"
+)
+
+func TestNv16Migration(t *testing.T) {
+	ctx := context.Background()
+	bs := cbor.NewMemCborStore()
+	adtStore := adt.WrapStore(ctx, bs)
+
+	startRoot := makeInputTree(ctx, t, adtStore)
+
+	oldStateTree, err := migration.LoadTree(adtStore, startRoot)
+	require.NoError(t, err)
+
+	oldSystemActor, found, err := oldStateTree.GetActor(builtin.SystemActorAddr)
+	require.NoError(t, err)
+	require.True(t, found, "system actor not found")
+
+	var oldSystemState system9.State
+	err = adtStore.Get(ctx, oldSystemActor.Head, &oldSystemState)
+	require.NoError(t, err)
+
+	oldManifestDataCid := oldSystemState.BuiltinActors
+	oldManifest := manifest.Manifest{
+		Version: 1,
+		Data:    oldManifestDataCid,
+	}
+	require.NoError(t, oldManifest.Load(ctx, adtStore), "failed to load old manifest")
+
+	newManifestCid, newManifestDataCid := makeTestManifest(t, adtStore, "fil/9/")
+	log := migration.TestLogger{TB: t}
+
+	cache := migration.NewMemMigrationCache()
+	_, err = migration.MigrateStateTree(ctx, adtStore, newManifestCid, startRoot, 200, migration.Config{MaxWorkers: 1}, log, cache)
+	require.NoError(t, err)
+
+	cacheRoot, err := migration.MigrateStateTree(ctx, adtStore, newManifestCid, startRoot, 200, migration.Config{MaxWorkers: 1}, log, cache)
+	require.NoError(t, err)
+
+	noCacheRoot, err := migration.MigrateStateTree(ctx, adtStore, newManifestCid, startRoot, 200, migration.Config{MaxWorkers: 1}, log, migration.NewMemMigrationCache())
+	require.NoError(t, err)
+	require.True(t, cacheRoot.Equals(noCacheRoot))
+
+	// TODO: check all the CIDs
+
+	// check that the system actor state was correctly updated
+
+	newStateTree, err := migration.LoadTree(adtStore, cacheRoot)
+	require.NoError(t, err)
+
+	newSystemActor, found, err := newStateTree.GetActor(builtin.SystemActorAddr)
+	require.NoError(t, err)
+	require.True(t, found, "system actor not found")
+
+	var newSystemState system9.State
+	err = adtStore.Get(ctx, newSystemActor.Head, &newSystemState)
+	require.NoError(t, err)
+	require.Equal(t, newManifestDataCid, newSystemState.BuiltinActors)
```

Comment:
We should leave a comment on this test to change this to check the states of the actors that we're changing in this upgrade (or maybe just set something up to check them all).

Reply:
Yessir